packages/arb-validator-core/gobridge/messagesChallenge.go

```diff
@@ -0,0 +1,391 @@
+	"github.com/offchainlabs/arbitrum/packages/arb-validator-core/arbbridge"
+)
+
+type messagesChallenge struct {
+	*bisectionChallenge
+}
+
+func newMessagesChallenge(address common.Address, client *GoArbAuthClient) (*messagesChallenge, error) {
+	bisectionChallenge, err := newBisectionChallenge(address, client) //, auth??
+	if err != nil {
+		return nil, err
+	}
+	vm := &messagesChallenge{bisectionChallenge: bisectionChallenge}
+	return vm, err
+}
+
+func (c *messagesChallenge) Bisect(
+	ctx context.Context,
+	chainHashes []common.Hash,
+	segmentHashes []common.Hash,
+	chainLength *big.Int,
+) error {
+
+	bisectionCount := len(chainHashes) - 1
+	if bisectionCount+1 != len(segmentHashes) {
+		return errors.New("Bisect Incorrect previous state - bisection count")
+	}
+
+	msgHash := valprotocol.MessageChallengeDataHash(chainHashes[0], chainHashes[bisectionCount], segmentHashes[0], segmentHashes[bisectionCount], chainLength)
+	if !c.client.GoEthClient.challenges[c.contractAddress].challengerDataHash.Equals(msgHash) {
+		return errors.New("Bisect Incorrect previous state msgHash")
+	}
+
+	hashes := make([][32]byte, 0, bisectionCount)
+	hashes = append(hashes, valprotocol.MessageChallengeDataHash(
+		chainHashes[0],
+		chainHashes[1],
+		segmentHashes[0],
+		segmentHashes[1],
+		new(big.Int).Add(new(big.Int).Div(chainLength, big.NewInt(int64(bisectionCount))), new(big.Int).Mod(chainLength, big.NewInt(int64(bisectionCount)))),
+	))
+	for i := 1; i < bisectionCount; i++ {
+		hashes = append(hashes, valprotocol.MessageChallengeDataHash(
+			chainHashes[i],
+			chainHashes[i+1],
+			segmentHashes[i],
+			segmentHashes[i+1],
+			new(big.Int).Div(chainLength, big.NewInt(int64(bisectionCount)))))
+	}
+
+	c.commitToSegment(hashes)
+	c.asserterResponded()
+
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.MessagesBisectionEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+			ChainHashes:   chainHashes,
+			SegmentHashes: segmentHashes,
+			TotalLength:   chainLength,
+			Deadline:      c.client.GoEthClient.challenges[c.contractAddress].deadline,
+		},
+	})
+
+	return nil
+}
+
+func (c *messagesChallenge) OneStepProofTransactionMessage(
+	ctx context.Context,
+	lowerHashA common.Hash,
+	lowerHashB common.Hash,
+	msg message.DeliveredTransaction,
+) error {
+	messageHash := msg.CommitmentHash()
+	arbMessageHash := message.DeliveredValue(msg).Hash()
+
+	// oneStepProof
+	if !c.challenge.challengeData.challengerDataHash.Equals(hashing.SoliditySHA3(
+		lowerHashA,
+		lowerHashB,
+		messageHash,
+		arbMessageHash,
+	)) {
+		return errors.New("OneStepProofTransactionMessage Incorrect previous state")
+	}
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.OneStepProofEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+		},
+	})
+	// TODO: challenge resolution
```

Comment:
The challenge currently doesn't seem to call the rollup chain to tell it the challenge is resolved.

Reply:
implemented resolveChallenge